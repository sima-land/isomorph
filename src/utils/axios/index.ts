export { type Sauce, type SauceResponse, sauce, formatResultInfo } from './sauce';
export { applyAxiosDefaults } from './utils';
export {
  type SharedData,
  type DoneSharedData,
  type FailSharedData,
  type LogMiddlewareHandler,
  type LogMiddlewareHandlerFactory,
  type LogMiddlewareHandlerInit,
  logMiddleware,
} from './middleware/log';
