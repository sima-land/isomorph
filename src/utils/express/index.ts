export { healthCheck } from './handler/health-check';
export { composeMiddleware } from './utils';
