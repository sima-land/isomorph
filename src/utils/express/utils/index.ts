import type express from 'express';

/**
 * Объединяет промежуточные слои в один.
 * @param list Промежуточные слои.
 * @return Промежуточный слой.
 */
export function composeMiddleware(list: express.Handler[]): express.Handler {
  return list.reduce((a, b) => (req, res, next) => {
    a(req, res, err => {
      if (err) {
        return next(err);
      }

      b(req, res, next);
    });
  });
}
