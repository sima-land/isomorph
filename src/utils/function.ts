import { identity } from 'lodash';

interface SafetyDoneInfo<T> {
  ok: true;
  result: T;
}

interface SafetyFailInfo {
  ok: false;
  error: unknown;
}

export type SafetyInfo<T> = SafetyDoneInfo<T> | SafetyFailInfo;

interface SafetyFormatter<T, F> {
  (info: SafetyInfo<T>): F | Promise<F>;
}

/**
 * Возвращает функцию, результатом которой является promise, который никогда не переходит в состояние rejected.
 * При перехвате исключения результат будет содержать ошибку.
 * @param fn Оборачиваемая функция.
 * @return Функция.
 */
export function safetyAsync<T, P extends any[]>(
  fn: (...args: P) => Promise<T>,
): (...args: P) => Promise<SafetyInfo<T>>;

/**
 * Возвращает функцию, результатом которой является promise, который никогда не переходит в состояние rejected.
 * При перехвате исключения результат будет содержать ошибку.
 * @param fn Оборачиваемая функция.
 * @param format Получив базовую информацию об ответе должна вернуть отформатированную информацию.
 * @return Функция.
 */
export function safetyAsync<T, P extends any[], F>(
  fn: (...args: P) => Promise<T>,
  format: SafetyFormatter<T, F>,
): (...args: P) => Promise<F>;

/**
 * Возвращает функцию, результатом которой является promise, который никогда не переходит в состояние rejected.
 * При перехвате исключения результат будет содержать ошибку.
 * @param fn Оборачиваемая функция.
 * @param format Получив базовую информацию об ответе должна вернуть отформатированную информацию.
 * @return Функция.
 */
export function safetyAsync<T, P extends any[], F>(
  fn: (...args: P) => Promise<T>,
  format: SafetyFormatter<T, F> = identity,
): (...args: P) => Promise<F> {
  return async function safetyWrapper(...args: P) {
    try {
      return format({ ok: true, result: await fn(...args) });
    } catch (error) {
      return format({ ok: false, error });
    }
  };
}
