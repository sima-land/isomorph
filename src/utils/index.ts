export { safetyAsync, type SafetyInfo } from './function';
export { toMilliseconds } from './math';
