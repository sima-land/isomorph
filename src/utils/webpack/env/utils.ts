import { EnvPluginOptions } from './types';
import fs from 'node:fs';
import path from 'node:path';
import dotenv from 'dotenv';

/**
 * Формирует объект с переменными среды с учетом .env файлов.
 * @param options Опции.
 * @return Объект с переменными среды.
 */
export function defineEnv(
  options: Pick<EnvPluginOptions, 'dotenvUsage'> & { onError?: (error: unknown) => void },
): Record<string, string | undefined> {
  const result = { ...process.env };

  if (options.dotenvUsage && process.env.NODE_ENV) {
    try {
      const dotenvPath = path.join(process.cwd(), `.env.${process.env.NODE_ENV}`);

      if (fs.existsSync(dotenvPath)) {
        const parsed = dotenv.parse(fs.readFileSync(dotenvPath, 'utf-8'));

        // @todo дёрнуть бы эту логику из пакета dotenv, https://github.com/motdotla/dotenv/issues/690
        for (const key of Object.keys(parsed)) {
          if (!Object.prototype.hasOwnProperty.call(result, key)) {
            result[key] = parsed[key];
          }
        }
      }
    } catch (error) {
      options.onError?.(error);
    }
  }

  return result;
}

/**
 * Получив таблицу "название переменной среды >> значение переменной среды" вернёт определение для DefinePlugin.
 * @param input Таблица переменных.
 * @return Определение для DefinePlugin.
 */
export function asEnvVariables(input: Record<string, string | undefined>): Record<string, string> {
  const result: Record<string, string> = {};

  for (const [key, value] of Object.entries(input)) {
    if (key && typeof value === 'string') {
      result[`process.env.${key}`] = JSON.stringify(value);
    }
  }

  return result;
}
