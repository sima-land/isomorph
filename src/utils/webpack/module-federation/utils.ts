import type { OriginalShared, Shared, SharedObject } from './types';

/**
 * Возвращает скрипт инициализации удаленного модуля.
 * @internal
 * @param params Параметры.
 * @param params.serviceName Имя сервиса, содержащего удаленные модули.
 * @param params.containersGlobalKey Глобальная переменная, в которую добавляются удаленные контейнеры.
 * @param params.remoteEntriesGlobalKey Глобальная переменная, в которой хранятся пути к удаленным точкам входа.
 * @param params.remoteEntryPath Путь к remoteEntry, если его не надо получать в runtime.
 * @return Скрипт инициализации.
 */
export function createExternalConfig({
  serviceName,
  containersGlobalKey,
  remoteEntriesGlobalKey,
  remoteEntryPath,
}: {
  serviceName: string;
  remoteEntriesGlobalKey: string;
  containersGlobalKey: string;
  remoteEntryPath?: string;
}) {
  return `promise new Promise((resolveProxy, rejectProxy) => {
  let installed = false;
  let shareScope;
  
  if (window["${remoteEntriesGlobalKey}"]) {
    const proxy = {
      get(request) {
        return installed
          ? window["${containersGlobalKey}"]["${serviceName}"].get(request)
          : new Promise((resolveRequest, rejectRequest) => {
              const scriptElement = document.createElement("script");

              scriptElement.onload = () => {
                scriptElement.remove();
                const container = window["${containersGlobalKey}"]["${serviceName}"];
                try {
                  container.init(shareScope);
                  installed = true;
                } catch (e) {}
                resolveRequest(container.get(request));
              };
              
              scriptElement.onerror = () => {
                scriptElement.remove();
                rejectRequest(
                  new Error('Failed loading remoteEntry for "${serviceName}".')
                );
              };
              
              scriptElement.src = ${
                remoteEntryPath
                  ? `'${remoteEntryPath}'`
                  : `window['${remoteEntriesGlobalKey}']['${serviceName}']`
              };
              
              scriptElement.async = true;
              document.head.append(scriptElement);
            });
      },
      init(scope) {
        shareScope = scope;
      },
    };

    resolveProxy(proxy);
  } else {
    rejectProxy(
      new ReferenceError('Object "${remoteEntriesGlobalKey}" unavailable.')
    );
  }
});`;
}

/**
 * Объединяет две коллекции модулей, переопределяя в base совпадающие.
 * @param base Базовая коллекция модулей.
 * @param expanding Дополнительная коллекция модулей.
 * @return Объект модулей.
 */
export function mergeModules(base: Shared, expanding: Shared): OriginalShared {
  const merged = { ...toObject(base), ...toObject(expanding) };

  return Object.fromEntries(
    Object.entries(merged).filter(
      (pair): pair is [string, Exclude<typeof pair[1], false>] => pair[1] !== false,
    ),
  );
}

/**
 * Преобразует переданную коллекцию в объект.
 * @param shared Коллекция модулей.
 * @return Объект модулей.
 */
function toObject(shared: Shared): SharedObject {
  return wrapInArray(shared).reduce((result: SharedObject, item) => {
    if (typeof item === 'string') {
      result[item] = {}; // Валидное определение модуля (все значения из дефолтов).
      return result;
    } else {
      return { ...result, ...item };
    }
  }, {});
}

/**
 * Оборачивает элемент в массив, если он не является массивом.
 * @param original Элемент.
 * @return Массив.
 */
function wrapInArray<T>(original: T | T[]): T[] {
  return Array.isArray(original) ? original : [original];
}
