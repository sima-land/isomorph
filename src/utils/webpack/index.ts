import { pick, pickBy } from 'lodash';
import webpack from 'webpack';
import dotenv from 'dotenv';
import path from 'path';
import { Env } from '@humanwhocodes/env';

/**
 * Плагин для зашивания переменных среды в сборки.
 * В браузерную сборку зашьёт только переменные с префиксом "PUBLIC_".
 * ВАЖНО: все зашиваемые переменные будут доступны через глобальный объект __ISOMORPH_ENV__.
 */
export class EnvPlugin {
  options: {
    additional: string[];
    define?: Record<string, string>;
  };

  /**
   * @param options Опции.
   */
  constructor(options: { additional?: string[]; define?: Record<string, string> } = {}) {
    this.options = {
      additional: options.additional || [],
      define: options.define,
    };
  }

  /**
   * @param compiler Компилятор.
   */
  apply(compiler: webpack.Compiler) {
    const { NODE_ENV } = new Env().required;

    dotenv.config({ path: path.join(process.cwd(), `.env.${NODE_ENV}`) });

    const target = Array.isArray(compiler.options.target)
      ? compiler.options.target
      : [compiler.options.target];

    // browser
    if (target.includes('web')) {
      const valuesWeb: Record<string, string | undefined> = {
        ...this.options.define,
        ...pick(process.env, [...this.options.additional, 'NODE_ENV']),
        ...pickBy(process.env, (_, key) => key.startsWith('PUBLIC_')),
      };

      new webpack.DefinePlugin({ __ISOMORPH_ENV__: JSON.stringify(valuesWeb) }).apply(compiler);
      new webpack.DefinePlugin(
        Object.entries(valuesWeb).reduce<Record<string, string>>((acc, [key, value]) => {
          acc[`process.env.${key}`] = JSON.stringify(value);
          return acc;
        }, {}),
      ).apply(compiler);
    }

    // node
    if (!target.includes('web')) {
      const valuesNode: Record<string, string | undefined> = {
        ...this.options.define,
        ...pick(process.env, [...this.options.additional, 'NODE_ENV']),
      };

      new webpack.DefinePlugin({ __ISOMORPH_ENV__: JSON.stringify(valuesNode) }).apply(compiler);

      if (this.options.define) {
        new webpack.DefinePlugin(
          Object.entries(this.options.define).reduce<Record<string, string>>(
            (acc, [key, value]) => {
              acc[`process.env.${key}`] = JSON.stringify(value);
              return acc;
            },
            {},
          ),
        ).apply(compiler);
      }
    }
  }
}
