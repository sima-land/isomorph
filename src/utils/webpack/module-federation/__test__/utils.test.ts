import { createExternalConfig } from '../utils';

describe('createExternalConfig', () => {
  it('should match reference', () => {
    expect(
      createExternalConfig({
        serviceName: 'example',
        containersGlobalKey: 'container_key',
        remoteEntriesGlobalKey: 'entries_key',
      }),
    ).toMatchSnapshot();

    expect(
      createExternalConfig({
        serviceName: 'example',
        containersGlobalKey: 'container_key',
        remoteEntriesGlobalKey: 'entries_key',
        remoteEntryPath: 'path/to/cdn/example/1.0.0/remoteEntry.hash.js',
      }),
    ).toMatchSnapshot();
  });
});
