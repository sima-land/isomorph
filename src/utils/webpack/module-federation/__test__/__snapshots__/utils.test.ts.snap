// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createExternalConfig should match reference 1`] = `
"promise new Promise((resolveProxy, rejectProxy) => {
  let installed = false;
  let shareScope;
  
  if (window[\\"entries_key\\"]) {
    const proxy = {
      get(request) {
        return installed
          ? window[\\"container_key\\"][\\"example\\"].get(request)
          : new Promise((resolveRequest, rejectRequest) => {
              const scriptElement = document.createElement(\\"script\\");

              scriptElement.onload = () => {
                scriptElement.remove();
                const container = window[\\"container_key\\"][\\"example\\"];
                try {
                  container.init(shareScope);
                  installed = true;
                } catch (e) {}
                resolveRequest(container.get(request));
              };
              
              scriptElement.onerror = () => {
                scriptElement.remove();
                rejectRequest(
                  new Error('Failed loading remoteEntry for \\"example\\".')
                );
              };
              
              scriptElement.src = window['entries_key']['example'];
              
              scriptElement.async = true;
              document.head.append(scriptElement);
            });
      },
      init(scope) {
        shareScope = scope;
      },
    };

    resolveProxy(proxy);
  } else {
    rejectProxy(
      new ReferenceError('Object \\"entries_key\\" unavailable.')
    );
  }
});"
`;

exports[`createExternalConfig should match reference 2`] = `
"promise new Promise((resolveProxy, rejectProxy) => {
  let installed = false;
  let shareScope;
  
  if (window[\\"entries_key\\"]) {
    const proxy = {
      get(request) {
        return installed
          ? window[\\"container_key\\"][\\"example\\"].get(request)
          : new Promise((resolveRequest, rejectRequest) => {
              const scriptElement = document.createElement(\\"script\\");

              scriptElement.onload = () => {
                scriptElement.remove();
                const container = window[\\"container_key\\"][\\"example\\"];
                try {
                  container.init(shareScope);
                  installed = true;
                } catch (e) {}
                resolveRequest(container.get(request));
              };
              
              scriptElement.onerror = () => {
                scriptElement.remove();
                rejectRequest(
                  new Error('Failed loading remoteEntry for \\"example\\".')
                );
              };
              
              scriptElement.src = 'path/to/cdn/example/1.0.0/remoteEntry.hash.js';
              
              scriptElement.async = true;
              document.head.append(scriptElement);
            });
      },
      init(scope) {
        shareScope = scope;
      },
    };

    resolveProxy(proxy);
  } else {
    rejectProxy(
      new ReferenceError('Object \\"entries_key\\" unavailable.')
    );
  }
});"
`;
